lib/rbi/rewriters/merge_trees.rb

```diff
@@ -212,6 +213,169 @@ def replace_scope_header(left, right)
+
+          merge_comments(left, right) if left.is_a?(NodeWithComments) && right.is_a?(NodeWithComments)
+
+          case left
+          when Class
+            right = right #: as Class
+            left_superclass = lookup_type(name: left.superclass_name, referrer: left)
+            right_superclass = lookup_type(name: right.superclass_name, referrer: right)
+            left_superclass == right_superclass
+          when Struct
+            right = right #: as Struct
+            left.members == right.members && left.keyword_init == right.keyword_init
+          when Const
+            right = right #: as Const
+            left.name == right.name && left.value == right.value
+          when Attr, Method
+            right = right #: as Attr | Method
+            return false if left.is_a?(Method) && right.is_a?(Method) && left.params != right.params
+            return false if left.is_a?(Attr) && right.is_a?(Attr) && left.names != right.names
+
+            left_sigs = left.sigs.map { fully_qualify_sig(_1, referrer: left) }
+            right_sigs = right.sigs.map { fully_qualify_sig(_1, referrer: right) }
+            if left_sigs.empty? || right_sigs.empty? || left_sigs == right_sigs
+              right_sigs.each do |sig|
+                left_sigs << sig unless left_sigs.include?(sig)
+              end
+              left.sigs = left_sigs
+              true
+            else
+              false
+            end
+          when Mixin
+            right = right #: as Mixin
+            left_mixins = left.names.map { lookup_type(name: _1, referrer: left) }
+            right_mixins = right.names.map { lookup_type(name: _1, referrer: right) }
+            left_mixins == right_mixins
+          when Helper
+            # Do Helper names need to be resolved to types?
+            right = right #: as Helper
+            left.name == right.name
+          when Send
+            right = right #: as Send
+            left.method == right.method && left.args == right.args
+          when TStructField
+            right = right #: as TStructField
+            left.name == right.name && left.type == right.type && left.default == right.default
+          else
+            true
+          end
+        end
+
+        # Returns a node from the merge tree that corresponds to the given type name
+        # when referenced from the given referrer Node. The referrer can be
+        # in a different tree, but its scope chain names will be used to find
+        # the referent in this merge tree.
+        #: (name: String?, referrer: Node) -> (Scope | Const)?
+        def lookup_type(name:, referrer:)
+          return unless name
+
+          if name.start_with?("::")
+            referent = @index[name].last #: Node?
+            if referent.is_a?(Scope) || referent.is_a?(Const)
+              return referent
+            elsif referent
+              raise "Unexpected type #{referent} for #{name} with referrer #{referrer}"
```

Comment:
I think this should raise a new `RBI::Merge::Error` similar to rest of the library raising custom errors.